tests/tripal_chado/api/ChadoSchemaTest.php

```diff
@@ -0,0 +1,308 @@
+      $this->assertArrayHasKey(
+        $table_name,
+        $returned_tables,
+        t('The table, "!known", should exist in the returned tables list for version !version.',
+          array(':known' => $table_name, ':version' => $version))
+      );
+    }
+  }
+
+  /**
+   * Tests ChadoSchema->getTableSchema() method.
+   *
+   * @dataProvider chadoTableProvider
+   *
+   * @group api
+   * @group chado
+   * @group chado-schema
+   */
+  public function testGetTableSchema($version, $table_name) {
+
+    // Check: a schema is returned that matches what we expect.
+    $chado_schema = new \ChadoSchema($version);
+
+    $table_schema = $chado_schema->getTableSchema($table_name);
+
+    $this->assertNotEmpty(
+      $table_schema,
+      t('Returned schema for "!table" in chado v!version should not be empty.',
+        array('!table' => $table_name, '!version' => $version))
+    );
+
+    $this->assertArrayHasKey(
+      'fields',
+      $table_schema,
+      t('The schema array for "!table" should have columns listed in an "fields" array',
+        array('!table' => $table_name))
+    );
+
+    // Instead of asserting these keys exist. Lets assert that if they do exist,
+    // they match the expected format.
+
+    if (isset($table_schema['primary key'])) {
+      $this->assertTrue(is_array($table_schema['primary key']),
+	t('The primary key of the Tripal Schema definition for "!table" must be an array.',
+          array('!table' => $table_name)));
+  
+    }
+
+    $this->assertArrayHasKey(
+      'foreign keys',
+      $table_schema,
+      t('The schema array for "!table" should have foreign keys listed in an "foreign keys" array',
+        array('!table' => $table_name))
+    );
+
+  }
+
+  /**
+   * Tests ChadoSchema->getCustomTableSchema() method.
+   *
+   * @dataProvider knownCustomTableProvider
+   *
+   * @group api
+   * @group chado
+   * @group chado-schema
+   */
+  public function testGetCustomTableSchema($table_name) {
+
+    // Check: a schema is returned that matches what we expect.
+    $chado_schema = new \ChadoSchema();
+    $table_schema = $chado_schema->getCustomTableSchema($table_name);
+
+    $this->assertNotEmpty(
+      $table_schema,
+      t('Returned schema for "!table" in chado v!version should not be empty.',
+        array('!table' => $table_name, '!version' => $version))
+    );
+
+    $this->assertArrayHasKey(
+      'fields',
+      $table_schema,
+      t('The schema array for "!table" should have columns listed in an "fields" array',
+        array('!table' => $table_name))
+    );
+
+    // NOTE: Other then ensuring fields are set, we can't test further since all other
+    // keys are technically optional and these arrays are set by admins.
+
+  }
+
+  /**
+   * Tests ChadoSchema->getBaseTables() method.
+   *
+   * @dataProvider knownBaseTableProvider
+   *
+   * @group api
+   * @group chado
+   * @group chado-schema
+   */
+  public function testGetBaseTables($version, $table_name) {
```

Comment:
Risky test no assertions.  I think thats because `$table_name` in this function definition is meant to be `$known_tables`.  That variable is undefined so the foreach loop doesnt happen.)

Reply:
hmm, setting that causes the test to fail:

```

There were 2 failures:

1) Tests\tripal_chado\api\ChadoSchemaTest::testGetBaseTables with data set #0 ('1.2', array('organism', 'feature', 'stock', 'project', 'analysis', 'phylotree'))
The table, "!known", should exist in the returned base tables list for version !version.
Failed asserting that an array has the key 'organism'.

/Users/chet/UTK/tripal/sites/all/modules/tripal/tests/tripal_chado/api/ChadoSchemaTest.php:215

2) Tests\tripal_chado\api\ChadoSchemaTest::testGetBaseTables with data set #1 ('1.3', array('organism', 'feature', 'stock', 'project', 'analysis', 'phylotree'))
The table, "!known", should exist in the returned base tables list for version !version.
Failed asserting that an array has the key 'organism'.

/Users/chet/UTK/tripal/sites/all/modules/tripal/tests/tripal_chado/api/ChadoSchemaTest.php:215
```

Reply:
ok it fails because getBaseTables returns an array of tables, but the array isnt indexed by anything IE asserting that an arrayHasKey is always going to fail.  

Should the test change to loop through the array and see if it has the *value* of the table name?   Or should the function change to return the array indexed by table name?